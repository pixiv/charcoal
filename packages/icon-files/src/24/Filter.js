/** This file is auto generated. DO NOT EDIT BY HAND. */
export default '<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M10.5 13.468L3.574 6.4C2.335 5.133 3.232 3 5.005 3h13.99c1.773 0 2.67 2.133 1.43 3.4L13.5 13.467V19a3 3 0 01-3.002 3v-8.532z" fill="currentColor"/></svg>'
