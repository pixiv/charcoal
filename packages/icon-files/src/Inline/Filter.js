/** This file is auto generated. DO NOT EDIT BY HAND. */
export default '<svg width="10" height="12" viewBox="0 0 10 12" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M.099 1.567A1 1 0 011 1h8a1 1 0 01.78 1.625L6 7.35V10a2 2 0 01-2 2V7.35L.22 2.626a1 1 0 01-.121-1.058z" fill="currentColor"/></svg>'
