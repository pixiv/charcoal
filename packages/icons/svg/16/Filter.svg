<svg width="16" height="16" viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M3.099 3.567A1 1 0 014 3h8a1 1 0 01.78 1.625L9 9.35V12a2 2 0 01-2 2V9.35L3.22 4.626a1 1 0 01-.121-1.058z" fill="currentColor"/></svg>