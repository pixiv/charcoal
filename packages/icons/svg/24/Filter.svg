<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg"><path fill-rule="evenodd" clip-rule="evenodd" d="M16.679 16.679a7.5 7.5 0 10-9.357-9.357 7.5 7.5 0 109.357 9.357zM20 9.5a5.5 5.5 0 11-11 0 5.5 5.5 0 0111 0z" fill="currentColor"/></svg>