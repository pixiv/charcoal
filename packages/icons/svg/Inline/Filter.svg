<svg width="10" height="12" viewBox="0 0 10 12" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M0.0987061 1.56681C0.265226 1.22035 0.615604 1 1.00001 1H9.00001C9.38441 1 9.73479 1.22035 9.90131 1.56681C10.0678 1.91328 10.021 2.32453 9.78088 2.6247L6.00002 7.35077V10C6.00002 11.1046 5.10458 12 4.00002 12V7.35079L0.219141 2.6247C-0.0209954 2.32453 -0.0678138 1.91328 0.0987061 1.56681Z" fill="#858585"/>
</svg>
