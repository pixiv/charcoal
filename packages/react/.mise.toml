[tools]
pnpm = "10.17.1"
