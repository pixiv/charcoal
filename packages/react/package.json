{
  "name": "@charcoal-ui/react",
  "version": "4.0.0-beta.3",
  "license": "Apache-2.0",
  "main": "./dist/index.cjs.js",
  "module": "./dist/index.esm.js",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "require": "./dist/index.cjs.js",
      "import": "./dist/index.esm.js",
      "default": "./dist/index.esm.js"
    },
    "./dist/index.css": {
      "import": "./dist/index.css",
      "require": "./dist/index.css"
    }
  },
  "types": "./dist/index.d.ts",
  "sideEffects": false,
  "scripts": {
    "build": "run-p --print-label 'build:*'",
    "build:bundle": "FORCE_COLOR=1 tsup",
    "build:dts": "tsc --project tsconfig.build.json --pretty --emitDeclarationOnly",
    "typecheck": "tsc --project tsconfig.build.json --pretty --noEmit",
    "clean": "rimraf dist .tsbuildinfo"
  },
  "devDependencies": {
    "@charcoal-ui/esbuild-plugin-styled-components": "^4.0.0-beta.3",
    "@react-types/switch": "^3.1.2",
    "@storybook/addon-actions": "^7.4.1",
    "@storybook/addon-knobs": "^7.0.2",
    "@storybook/addons": "^7.4.1",
    "@storybook/api": "^7.4.1",
    "@storybook/components": "^7.4.1",
    "@storybook/core-events": "^7.4.1",
    "@storybook/react": "^7.4.1",
    "@storybook/react-webpack5": "^7.4.1",
    "@storybook/theming": "^7.4.1",
    "@testing-library/jest-dom": "^5.16.1",
    "@testing-library/react": "^12.1.2",
    "@testing-library/user-event": "^13.5.0",
    "@types/jest": "^27.4.0",
    "@types/jest-axe": "^3.5.3",
    "@types/react": "^17.0.38",
    "@types/react-dom": "^17.0.11",
    "@types/react-router-dom": "^5.3.3",
    "@types/styled-components": "^5.1.21",
    "@types/warning": "^3.0.0",
    "jest-axe": "^5.0.1",
    "jest-styled-components": "^7.0.8",
    "npm-run-all": "^4.1.5",
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "react-router-dom": "^6.2.1",
    "rimraf": "^3.0.2",
    "storybook": "^7.4.1",
    "styled-components": "^5.3.3",
    "tsup": "^6.5.0",
    "typescript": "^4.9.5"
  },
  "dependencies": {
    "@charcoal-ui/icons": "^4.0.0-beta.3",
    "@charcoal-ui/styled": "^4.0.0-beta.3",
    "@charcoal-ui/theme": "^4.0.0-beta.3",
    "@charcoal-ui/utils": "^4.0.0-beta.3",
    "@react-aria/button": "^3.9.1",
    "@react-aria/checkbox": "^3.13.0",
    "@react-aria/dialog": "^3.5.10",
    "@react-aria/focus": "^3.16.0",
    "@react-aria/link": "^3.6.5",
    "@react-aria/overlays": "^3.20.0",
    "@react-aria/radio": "^3.10.0",
    "@react-aria/ssr": "^3.9.1",
    "@react-aria/switch": "^3.6.0",
    "@react-aria/textfield": "^3.14.1",
    "@react-aria/utils": "^3.23.0",
    "@react-aria/visually-hidden": "^3.8.8",
    "polished": "^4.1.4",
    "react-spring": "^9.0.0",
    "react-stately": "^3.26.0",
    "warning": "^4.0.3"
  },
  "peerDependencies": {
    "react": ">=17.0.0",
    "styled-components": ">=5.1.1"
  },
  "files": [
    "src",
    "dist"
  ],
  "publishConfig": {
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/pixiv/charcoal.git",
    "directory": "packages/react"
  },
  "gitHead": "e1ece2e43901ae667afdd5c178040607d939dcd5"
}
