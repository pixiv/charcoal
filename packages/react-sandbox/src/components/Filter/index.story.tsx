import { action } from '@storybook/addon-actions'
import { boolean } from '@storybook/addon-knobs'
import React from 'react'
import {
  Link as RouterLink,
  MemoryRouter as Router,
  Route,
  useParams,
  useNavigate,
} from 'react-router-dom'
import Filter, { FilterButton, FilterLink } from '.'
import { ComponentAbstraction } from '@charcoal/react'

export default {
  title: 'Sandbox/Filter',
  component: Filter,
}

const makeUrl = (page: number) => `/${page}`

export function Default() {
  return (
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-expect-error TOOD: adapt to react-router@6  (props should be covariant not bivariant)
    <ComponentAbstraction components={{ Link: RouterLink }}>
      <Router
        initialEntries={Array.from({ length: 5 }).map((_, i) => makeUrl(i + 1))}
        initialIndex={0}
      >
        <Route path="/:page" element={<FilterButtons />} />
      </Router>
    </ComponentAbstraction>
  )
}

function FilterButtons() {
  const navigate = useNavigate()
  const params = useParams()
  const page = params.page !== undefined ? parseInt(params.page, 10) : 0

  return (
    <Filter>
      <FilterLink to={makeUrl(0)} active={page === 0}>
        新しい順
      </FilterLink>
      <FilterLink to={makeUrl(1)} active={page === 1}>
        古い順
      </FilterLink>
      <FilterButton active={page === 2} onClick={() => navigate('2')}>
        人気順
      </FilterButton>
      <FilterLink to={makeUrl(3)} active={page === 3}>
        これはリンク
      </FilterLink>
      <FilterButton active={page === 4} onClick={() => navigate('4')}>
        これはボタン
      </FilterButton>
    </Filter>
  )
}

export function Button() {
  const active = boolean('Active', true)
  const reactive = boolean('Reactive', false)
  return (
    <FilterButton onClick={action('click')} active={active} reactive={reactive}>
      Label
    </FilterButton>
  )
}

export function Link() {
  const active = boolean('Active', true)
  const reactive = boolean('Reactive', false)
  return (
    <FilterLink to="#" active={active} reactive={reactive}>
      Label
    </FilterLink>
  )
}
