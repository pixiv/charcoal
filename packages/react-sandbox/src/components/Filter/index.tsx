import React from 'react'
import styled, { css } from 'styled-components'
import { useComponentAbstraction, LinkProps } from '@charcoal/react'
import { maxWidth } from '@charcoal/utils'

interface Props<T extends HTMLElement> {
  active?: boolean
  hover?: boolean
  reactive?: boolean
  children: React.ReactNode
  onClick?: React.MouseEventHandler<T>
}

interface FilterIconButtonProps extends Props<HTMLButtonElement> {
  width?: number
  height?: number
}

export const FilterButton = React.forwardRef(function FilterButton(
  {
    onClick,
    children,
    active = false,
    hover,
    reactive = false,
  }: Props<HTMLButtonElement>,
  ref: React.Ref<HTMLButtonElement>
) {
  return (
    <ButtonLike
      active={active}
      reactive={reactive}
      hover={hover}
      onClick={active && !reactive ? undefined : onClick}
      ref={ref}
    >
      {children}
    </ButtonLike>
  )
})

export const FilterIconButton = React.forwardRef(function FilterIconButton(
  {
    onClick,
    children,
    active = false,
    hover,
    reactive = false,
    width,
    height,
  }: FilterIconButtonProps,
  ref: React.Ref<HTMLButtonElement>
) {
  return (
    <StyledButtonLike
      active={active}
      reactive={reactive}
      hover={hover}
      onClick={active && !reactive ? undefined : onClick}
      ref={ref}
      buttonWidth={width}
      buttonHeight={height}
    >
      {children}
    </StyledButtonLike>
  )
})

export const FilterLink = React.forwardRef(function FilterLink(
  {
    onClick,
    children,
    active = false,
    hover,
    reactive = false,
    ...props
  }: Props<HTMLAnchorElement> & LinkProps,
  ref: React.Ref<HTMLAnchorElement>
) {
  const { Link } = useComponentAbstraction()
  if (active && !reactive) {
    return (
      <PlainElement active hover={hover} ref={ref}>
        {children}
      </PlainElement>
    )
  } else {
    return (
      <Link {...props} onClick={onClick}>
        <PlainElement
          active={active}
          reactive={reactive}
          hover={hover}
          ref={ref}
        >
          {children}
        </PlainElement>
      </Link>
    )
  }
})

interface ButtonCssProps {
  active?: boolean
  hover?: boolean
  reactive?: boolean
}

const buttonCss = css`
  display: block;
  outline: none;
  border: none;
  padding: 9px 24px;
  font-size: 14px;
  line-height: 22px;
  font-weight: bold;
  border-radius: /* absurd radius, but ensures rounded corners */ 2000px;
  transition: color 0.2s;
  color: ${({ theme }) => theme.color.text3};
  cursor: pointer;
  user-select: none;
  background-color: transparent;

  &:hover {
    color: ${({ theme }) => theme.color.text2};
  }

  ${({ hover = false }: ButtonCssProps) =>
    hover &&
    css`
      color: ${({ theme }) => theme.color.text2};
    `}

  ${({ active = false }: ButtonCssProps) =>
    active &&
    css`
      background-color: ${({ theme }) => theme.color.surface3};
      color: ${({ theme }) => theme.color.text2};
    `}

  ${({ active = false, reactive = false }) =>
    active &&
    !reactive &&
    css`
      cursor: default;
    `}

  @media ${({ theme }) => maxWidth(theme.breakpoint.screen1)} {
    padding: 5px 16px;
  }
`

const padding0Css = css`
  padding: 0;

  @media ${({ theme }) => maxWidth(theme.breakpoint.screen1)} {
    padding: 0;
  }
`

const ButtonLike = styled.button`
  ${buttonCss}
`

const PlainElement = styled.span`
  ${buttonCss}
`

const StyledButtonLike = styled(ButtonLike)<{
  buttonWidth: number | undefined
  buttonHeight: number | undefined
}>`
  ${padding0Css};
  ${(p) => p.buttonWidth !== undefined && `width: ${p.buttonWidth}px;`}
  ${(p) => p.buttonHeight !== undefined && `height: ${p.buttonHeight}px;`}
`

const Filter = styled.div`
  display: flex;
`

export default Filter
