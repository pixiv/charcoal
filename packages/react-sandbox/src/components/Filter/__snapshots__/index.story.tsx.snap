// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Sandbox/Filter Button 1`] = `
.c0 {
  display: block;
  outline: none;
  border: none;
  padding: 9px 24px;
  font-size: 14px;
  line-height: 22px;
  font-weight: bold;
  border-radius: 2000px;
  -webkit-transition: color 0.2s;
  transition: color 0.2s;
  color: #858585;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background-color: transparent;
  background-color: rgba(0,0,0,0.04);
  color: #474747;
  cursor: default;
}

.c0:hover {
  color: #474747;
}

@media (max-width:743px) {
  .c0 {
    padding: 5px 16px;
  }
}

<div
  data-dark={false}
>
  <button
    className="c0"
  >
    Label
  </button>
</div>
`;

exports[`Storyshots Sandbox/Filter Default 1`] = `
.c3 {
  display: block;
  outline: none;
  border: none;
  padding: 9px 24px;
  font-size: 14px;
  line-height: 22px;
  font-weight: bold;
  border-radius: 2000px;
  -webkit-transition: color 0.2s;
  transition: color 0.2s;
  color: #858585;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background-color: transparent;
}

.c3:hover {
  color: #474747;
}

.c1 {
  display: block;
  outline: none;
  border: none;
  padding: 9px 24px;
  font-size: 14px;
  line-height: 22px;
  font-weight: bold;
  border-radius: 2000px;
  -webkit-transition: color 0.2s;
  transition: color 0.2s;
  color: #858585;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background-color: transparent;
}

.c1:hover {
  color: #474747;
}

.c2 {
  display: block;
  outline: none;
  border: none;
  padding: 9px 24px;
  font-size: 14px;
  line-height: 22px;
  font-weight: bold;
  border-radius: 2000px;
  -webkit-transition: color 0.2s;
  transition: color 0.2s;
  color: #858585;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background-color: transparent;
  background-color: rgba(0,0,0,0.04);
  color: #474747;
  cursor: default;
}

.c2:hover {
  color: #474747;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

@media (max-width:743px) {
  .c3 {
    padding: 5px 16px;
  }
}

@media (max-width:743px) {
  .c1 {
    padding: 5px 16px;
  }
}

@media (max-width:743px) {
  .c2 {
    padding: 5px 16px;
  }
}

<div
  data-dark={false}
>
  <div
    className="c0"
  >
    <a
      href="/0"
      onClick={[Function]}
    >
      <span
        className="c1"
      >
        新しい順
      </span>
    </a>
    <span
      className="c2"
    >
      古い順
    </span>
    <button
      className="c3"
      onClick={[Function]}
    >
      人気順
    </button>
    <a
      href="/3"
      onClick={[Function]}
    >
      <span
        className="c1"
      >
        これはリンク
      </span>
    </a>
    <button
      className="c3"
      onClick={[Function]}
    >
      これはボタン
    </button>
  </div>
</div>
`;

exports[`Storyshots Sandbox/Filter Link 1`] = `
.c0 {
  display: block;
  outline: none;
  border: none;
  padding: 9px 24px;
  font-size: 14px;
  line-height: 22px;
  font-weight: bold;
  border-radius: 2000px;
  -webkit-transition: color 0.2s;
  transition: color 0.2s;
  color: #858585;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background-color: transparent;
  background-color: rgba(0,0,0,0.04);
  color: #474747;
  cursor: default;
}

.c0:hover {
  color: #474747;
}

@media (max-width:743px) {
  .c0 {
    padding: 5px 16px;
  }
}

<div
  data-dark={false}
>
  <span
    className="c0"
  >
    Label
  </span>
</div>
`;
